ionic-team/stencil
# feat(docs): enrich type information for docs-json Output Target

This PR adds significant functionality to the `docs-json` output target for documenting the types used in a Stencil project.

In particular, it adds a new property to the json blob emitted by `docs-json` called `typeLibrary` which includes extensive information about the types used in the public APIs of all the components included in a Stencil project.

Additionally, it adds an option to the `docs-json` output target, called `supplementalPublicTypes`, which accepts a filepath pointing to a file with, well, supplemental public types in it 😄 these are basically types which the developers for a Stencil project would like to include in their documentation but which are not used in a public API (e.g. a `@Prop`, an `@Event`, etc). The indicated file should be a typescript file which exports types of interest, along the lines of the [`interface.d.ts` file in Framework](https://github.com/ionic-team/ionic-framework/blob/main/core/src/interface.d.ts). All types and interfaces exported from this file will also be documented in the type library, preparing them for easy inclusion in autogenerated documentation or whatever other downstream use-case.



## Testing

To test this I'd suggest trying it out in Framework. Do this:

1. clone and build this branch
2. install it in framework
3. change the configuration of the `docs-json` output target in Framework to specify [`src/interface.d.ts`](https://github.com/ionic-team/ionic-framework/blob/main/core/src/interface.d.ts) for the `supplementalPublicTypes` property

Then do `npm run build` and look at the docs-json output in `docs/core.json`. Make sure that _all_ of the types in `src/interface.d.ts` are present, that the output makes sense, etc. A tool like `jq` might be helpful for this.
